src/L2VoteAggregator.sol

```diff
@@ -0,0 +1,142 @@
+  L2GovernorMetadata immutable GOVERNOR_METADATA;
+
+  /// @notice The contract that handles fetch the L1 block on the L2.
+  IL1Block immutable L1_BLOCK;
+
+  /// @dev Thrown when an address has no voting weight on a proposal.
+  error NoWeight();
+
+  /// @dev Thrown when an address has already voted.
+  error AlreadyVoted();
+
+  /// @dev Thrown when an invalid vote is cast.
+  error InvalidVoteType();
+
+  /// @dev Thrown when proposal is inactive.
+  error ProposalInactive();
+
+  /// @dev The voting options corresponding to those used in the Governor.
+  enum VoteType {
+    Against,
+    For,
+    Abstain
+  }
+
+  /// @dev Data structure to store vote preferences expressed by depositors.
+  // TODO: Does it matter if we use a uint128 vs a uint256?
+  struct ProposalVote {
+    uint128 against;
+    uint128 inFavor;
+    uint128 abstain;
+  }
+
+  /// @notice A mapping of proposal to a mapping of voter address to boolean indicating whether a
+  /// voter has voted or not.
+  mapping(uint256 => mapping(address => bool)) private _proposalVotersHasVoted;
+
+  /// @notice A mapping of proposal id to proposal vote totals.
+  mapping(uint256 => ProposalVote) public proposalVotes;
+
+  /// @param _votingToken The token used to vote on proposals.
+  /// @param _core The Wormhole contract to bridge messages to L1.
+  /// @param _governorMetadata The `GovernorMetadata` contract that provides proposal information.
+  /// @param l1BlockAddress The address of the L1Block contract.
+  constructor(
+    address _votingToken,
+    address _core,
+    address _governorMetadata,
+    address l1BlockAddress
+  ) {
+    VOTING_TOKEN = ERC20Votes(_votingToken);
+    CORE_BRIDGE = IWormhole(_core);
+    GOVERNOR_METADATA = L2GovernorMetadata(_governorMetadata);
+    L1_BLOCK = IL1Block(l1BlockAddress);
+  }
+
+  /// @notice Where a user can express their vote based on their L2 token voting power.
+  /// @param proposalId The id of the proposal to vote on.
+  /// @param support The type of vote to cast.
+  function castVote(uint256 proposalId, uint8 support) public returns (uint256 balance) {
+    if (!proposalVoteActive(proposalId)) revert ProposalInactive();
+    if (_proposalVotersHasVoted[proposalId][msg.sender]) revert AlreadyVoted();
+    _proposalVotersHasVoted[proposalId][msg.sender] = true;
+
+    L2GovernorMetadata.Proposal memory proposal = GOVERNOR_METADATA.getProposal(proposalId);
+    uint256 weight = VOTING_TOKEN.getPastVotes(msg.sender, proposal.voteStart);
+    if (weight == 0) revert NoWeight();
+
+    if (support == uint8(VoteType.Against)) {
+      proposalVotes[proposalId].against += SafeCast.toUint128(weight);
+    } else if (support == uint8(VoteType.For)) {
+      proposalVotes[proposalId].inFavor += SafeCast.toUint128(weight);
+    } else if (support == uint8(VoteType.Abstain)) {
+      proposalVotes[proposalId].abstain += SafeCast.toUint128(weight);
+    } else {
+      revert InvalidVoteType();
+    }
+    return weight;
```

Comment:
We should probably emit an event (matching the Governor contract) about the vote that has been cast.

Reply:
Good, idea the original Governor event has a reason param. Should we add reason to the `castVote` Governor signature?

Reply:
https://github.com/ScopeLift/l2-flexible-voting/pull/9/commits/c43aae933da93067a7cd415b24e8c75d66eaeb48